api/lib/opentelemetry/trace/span.rb

```diff
@@ -76,29 +76,38 @@ def set_attribute(key, value)
       # @return [self] returns itself
       def add_event(name, attrs = nil)
         raise ArgumentError if name.nil?
-        raise ArgumentError unless valid_attributes?(attrs)
+        raise ArgumentError unless Internal.valid_attributes?(attrs)
 
         self
       end
 
-      # Adds a link to another Span from this Span. Linked Span can be from the
-      # same or different trace. See Links description.
+      # Adds a link to another Span from this Span. The linked Span can be from
+      # the same or different trace. See {Link} for a description.
       #
-      # @param [SpanContext, Link] span_context_or_link SpanContext of the Span
-      #   to link to or a Link which allows allows instrumentation to supply a
-      #   lazily calculated Link.
+      # @param [SpanContext] span_context SpanContext of the Span to link with
+      #   Span
       # @param [Hash<String, Object>] attrs Map of attributes associated with
       #   this link. Attributes are key:value pairs where key is a string and
-      #   value is one of string, boolean or numeric type.
+      #   value is one of string, boolean or numeric type. If both attributes
+      #   and a link_formatter are provided, an ArgumentError will be raised.
+      #
+      # @yieldreturn [optional Hash<String, Object>] attribute_formatter A
+      #   callable that returns a hash of attributes for this link. Will be
+      #   called lazily and its return value will be frozen when attributes are
+      #   first accessed. If both attributes and a link_formatter are provided,
+      #   an ArgumentError will be raised.
       #
       # @return [self] returns itself
-      def add_link(span_context_or_link, attrs = nil)
-        raise ArgumentError if span_context_or_link.nil?
-        raise ArgumentError unless span_context_or_link.instance_of?(SpanContext) || attrs.nil? || attrs.empty?
-        raise ArgumentError unless valid_attributes?(attrs)
+      #
+      # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+      def add_link(span_context_or_link, attrs = nil, &attribute_formatter)
+        raise ArgumentError unless span_context_or_link.instance_of?(SpanContext) || span_context_or_link.is_a?(Link)
+        raise ArgumentError if span_context_or_link.instance_of?(SpanContext) && attrs.nil? && attribute_formatter.nil?
```

Comment:
I don't think this check is correct. It's OK to have a `SpanContext` without any attributes.